### PR TITLE
Create ModularArithmetic.cpp

### PR DESCRIPTION
**Pull request by rutuja2229**
issue -  #399 
Performing operations on larger input constraints of 10^50000, with given mod value, coded by GSSoC2021 participant.

### CODE REVIEW

**Comment by rutuja2229** [comment]
@tarun26091999  Can u plz check and merge this pull request?

**Review by tarun26091999 — CHANGES_REQUESTED** [review]
Problem explanation has to be at the top. Complexities to be specified at the end. At least 2 examples to be provided. Also provide link of your code in the appropriate folder.

**Comment by rutuja2229** [comment]
"Also provide the link of your code in the appropriate folder", can u plz help with this?

**Comment by rutuja2229** [comment]
@tarun26091999 plz check again.

File: Code/C++/ModularArithmetic.cpp [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+   From this, as of the problem, M is prime, express A^B mod M as follows: 
+          A^B mod M = ( A^(M-1) * A^(M-1) *.......* A^(M-1) * A^(x) ) mod M          
+   Where x is B mod M-1 and A ^ (M-1) continues B/(M-1) times
+   Now, from Fermat’s Little Theorem,  
+           A ^ (M-1) mod M = 1.
+           A^B mod M = ( 1 * 1 * ....... * 1 * A^(x) ) mod M
+   Hence mod B with M-1 to reduce the number to a smaller one and then use power() method to compute (a^b)%m. 
+            
+   
+TIME COMPLEXITY -
```
Comment (OjusWiZard) [inline Code/C++/ModularArithmetic.cpp:25]: Add time complexity below the code.

**Review by OjusWiZard — CHANGES_REQUESTED** [review]
Maintain the indentation of 4 spaces throughout the code. It is getting staggered in some lines.

File: Code/C++/ModularArithmetic.cpp (hunk 1 above)
Comment (rutuja2229) [reply Code/C++/ModularArithmetic.cpp]: @OjusWiZard  Completed the requested changes, please review this again.

**Comment by rutuja2229** [comment]
@OjusWiZard Completed the requested changes, please review this again.

File: Algorithm/Number_theory/readme.md [hunk 2]
```diff
@@ -39,4 +39,4 @@
 # Questions related to Number Theory
 
 - GCD of two number using Euclidean Algorithm ----> [Java](/Code/Java/GCD.java)
-- [Perform Basic operations like Addition,Subtraction,Multiplication,Division,Exponentiation on 10^500000 constraints](Code/C++/ModularArithmetic.cpp)
+- Perform Basic operations like Addition,Subtraction,Multiplication,Division,Exponentiation on 10^500000 constraints ----> [C++](/Code/C++/ModularArithmetic.cpp)
```
Comment (OjusWiZard) [inline Algorithm/Number_theory/readme.md:42]: It's better than before but try replacing it with some smaller title.
No need to specify all the operations in the title. Just say Modular Arithmetical Operations for Unsupported Large Integers.

**Review by OjusWiZard — CHANGES_REQUESTED** [review]

File: Algorithm/Number_theory/readme.md (hunk 2 above)
Comment (rutuja2229) [reply Algorithm/Number_theory/readme.md:42]: okay, Sorry for the conflicts.

**Review by OjusWiZard — APPROVED** [review]

**Comment by rutuja2229** [comment]
When will my PR get merge @OjusWiZard ?

File: Code/C++/ModularArithmetic.cpp [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+            b=modm(y);
+            cout<<((a-b + m) % m)<<"\n"; // to handle the underflow, we have to add the mod value and then calculate the modulo operation
+        }
+        else if(si=="*"){
+           a=modm(x);
+            b=modm(y);
+            cout<<((a*b) % m)<<"\n";
+        }
+        else if(si=="^"){
+             ll res = 1;
```
Comment (OjusWiZard) [inline Code/C++/ModularArithmetic.cpp:113]: Correct the indentation.

File: Code/C++/ModularArithmetic.cpp [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+            while (b > 0)  // calculating a^b after performing modular operation on them
+            {  
+                if (b & 1)  
+                    res = (res*a) % m;  
+                a = (a*a) % m;
+                b = b>>1; //Similar to  b=b/2
+                  
+            }
+            cout<<(res%m)<<"\n";
+          }
```
Comment (OjusWiZard) [inline Code/C++/ModularArithmetic.cpp:132]: Correct the indentation.

File: Code/C++/ModularArithmetic.cpp [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+            b=modm(y);
+            ll g = gcd1(a,b);
+            if(g!=1){
+                a=a/g;
+                b=b/g;
+            }
+            ll inv = inverse(b, m); // see the explaination above 
+            if (inv == -1) 
+               cout <<-1<<"\n"; 
+            else
```
Comment (OjusWiZard) [inline Code/C++/ModularArithmetic.cpp:145]: Correct the indentation of if and else.

File: Code/C++/ModularArithmetic.cpp [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+               cout<< (a * inv) % m <<"\n";
+        }
+        
+        
+
+
```
Comment (OjusWiZard) [inline Code/C++/ModularArithmetic.cpp:151]: Remove extra white space.

**Review by OjusWiZard — CHANGES_REQUESTED** [review]
I would suggest you to use some code auto formatted to deal with so many formatting issues.

File: Code/C++/ModularArithmetic.cpp [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+        return b; 
+    } 
+  
+    ll x1, y1;  
+    ll g = gcd(b%a, a, &x1, &y1); 
+    *x = y1 - (b/a) * x1; 
+    *y = x1; 
+  
+    return g; 
+}
```
[the remark this answers is not in the document]
Comment (rutuja2229) [reply Code/C++/ModularArithmetic.cpp:67]: @OjusWiZard for performing inverse operation this gcd() function is being called.

File: Code/C++/ModularArithmetic.cpp (hunk 6 above)
Comment (rutuja2229) [reply Code/C++/ModularArithmetic.cpp:151]: @yasharth291 I am trying, due to exams I was inactive, apologizes for this. Please add this problem. I will try to update that.Please.
@OjusWiZard Please.

**Pull request closed without merging**